### DISCUSSION TITLE
Please help me with ops parameters and configurations.

### DISCUSSION THREAD

**Discussion by dzciemix** [dagster-io/dagster#14395] [Q&A] [answered]
Hello!

First of, thank you for the project :) It looks great!

I have problem understanding the dagster way of ops parameters and configurations. Lets say I have op and want run multiple ops with different parameters.

```
AConfig(Config):
    max: int

@op
def a_op(host: str, command: str):
    pass

@graph
def a_graph(config: AConfig):
    for host, command in [("a", "echo a "), ("b", "echo b")]:
        a_op(host, command + str(random(config.max)))
```

How can I do that? How can I make parameters semi-automated? Why can't I just run `a_op("h", "pwd")`?

**Comment by OwenKephart** [comment, marked as answer]
Hi @dzciemix ! The general pattern we recommend here is an [op factory](https://docs.dagster.io/concepts/ops-jobs-graphs/ops#op-factory). In your case, that'd look like:

```
def get_a_op(host: str, command: str):
    @op(name=f"a_op_{host}_{command}")
    def a_op():
        pass
    return a_op
    
 def get_a_graph(max: int):
     @graph(name=f"a_graph_{max}")
     def a_graph():
         for host, command in ...:
             get_a_op(host, command + str(random(max)))()
     return a_graph
```

Definitely a bit roundabout, but this mostly just comes down to the separation between what's known at definition time vs what's known at runtime.

The config can't impact the graph's structure, because Dagster needs to know the shape of the graph before it's configured. As for just running `a_op("h", "pwd")`, this is definitely something that we've considered implementing, but just hasn't been prioritized as there are other ways to achieve that functionality (like the factory approach)
